### ISSUE TITLE
(RMA)  Relative Moving Average

### ISSUE THREAD

**Issue by mario4tier** [TA-Lib/ta-lib#348]
Original request from @DelphiMan68:
https://github.com/TA-Lib/ta-lib/issues/41

---

## What it is

Wilder's smoothed moving average: an EMA whose smoothing factor is `1/n` instead of `2/(n+1)`,
SMA-seeded over the first `n` bars. J. Welles Wilder Jr., *New Concepts in Technical Trading
Systems* (1978) — the same smoothing already embedded inside the shipped `TA_ATR`, `TA_RSI`,
`TA_ADX`, `TA_PLUS_DM`. This exposes it as a standalone MA. Sold under five names for one object:
RMA (TradingView, pandas-ta), SMMA (MetaTrader), WildersSmoothing / WildersAverage (thinkorswim),
`wilders` (Tulip), WilderMA (Wealth-Lab).

## Formula

```
alpha = 1.0 / n
beta  = 1.0 - alpha

seed  (at bar n-1):   sum = 0; for i in [0 .. n-1] sum += x[i];   v = sum / n
recurrence (bar i>=n):                                            v = alpha*x[i] + beta*v
```

First valid output at bar `n-1`. Sequential seed accumulation, in input order.

**The arithmetic form is load-bearing.** Three algebraically identical forms are in published use:

| Form | Used by |
|---|---|
| `alpha*x + (1-alpha)*v` | **PROPOSED** — TradingView Pine `ta.rma`; pandas `ewm(adjust=False)` kernel; TA-Lib `TA_ATR`/`TA_NATR` after #338; TA-Lib `T3`/`ADOSC`/`MAMA`/`HT_*` (87 sites) |
| `(x - v)*alpha + v` | Tulip `wilders.c:55`; TA-Lib's `ema_private` (`input/ema/ema.c:117`) |
| `v *= n-1; v += x; v /= n` | MetaTrader, Wealth-Lab, thinkorswim; TA-Lib `TA_ATR` before #338 |

**Form 1, spelled exactly as shipped `TA_ATR` spells it** (`ta_codegen/input/atr/atr.c:77`):

```c
wBeta  = (double)(optInTimePeriod - 1) / (double)optInTimePeriod;
wAlpha = 1.0 - wBeta;                    /* derived FROM wBeta, never the reverse */
v      = wAlpha * x + wBeta * v;         /* generator emits fma(wBeta, v, wAlpha*x) */
```

**beta-first, not `alpha = 1.0/n`.** The two spellings give a different `alpha` double at 199982 of
the first 200000 periods, so this is the whole decision. MEASURED:

| | beta-first (shipped) | alpha-first |
|---|---|---|
| `RMA(TRANGE,n) == ATR(n)` memcmp | **holds** | 5.78e-15, 7% of values bit-equal |
| ATR vs the frozen v0.6.4 legacy samples, n=19 | 4.44e-16 | **4.885e-15 — RED against `LEGACY_TOL` 3e-15** |
| vs the form-2 oracle family (Tulip = ta4j = trading-signals WSMA) | 39–52% of tail bits | 5–10% |
| vs pandas / trading-signals RMA | 9.3% of bits, worst 4.70e-15 | 23.5%, worst 1.79e-15 |
| accuracy vs an exact rational recursion, n=2..120 | median 4.2e-16 | median 1.4e-15 |
| `wAlpha + wBeta == 1` in exact arithmetic | **all 200000 periods** | fails at 199982 |
| per-bar cost | identical | identical |

Two measurement traps, stated so they are not re-run wrong. **Model the step with a real `fma`** —
`a*x + b*v` in a language without one is a different library, and it is where "alpha-first is
bit-identical to pandas at 110/119 periods" comes from; with the shipped `fma` that is 1/119, and
bitwise pandas parity is unreachable under any spelling because pandas' comass round-trip is lossy
at 9 of 119 periods. And **ta4j `MMAIndicator` is form 2**, the same kernel as Tulip `wilders` and
trading-signals `WSMA` — group the arms by step form, not by alpha.

Alpha-first's one real gain is 2.5x closer to pandas. Driving `ewm` at **our** alpha instead of
`1/n` gains 4.1x (4.70e-15 -> 1.14e-15 vs real pandas 3.0.5) with no library change, and works only
for beta-first: `fl(wAlpha+wBeta) == 1.0` at every period makes pandas' renormalisation a no-op.

`n == 1` and `n == 2` are bit-identical under both spellings, so the period-1 leg cannot gate this.

**Rejected: form 2** (`prev + (x-prev)*alpha`). It satisfies both contested properties — DC-exact
structurally, and it carries the canonical `fl(1/n)` — and is *more* accurate (median 3.69e-16) and
matches the form-2 family at 87.2%. It costs **1.66x on the batch loop**, giving back 40% of
TA-Lib/ta-lib#338, and it silently breaks the period-1 identity off price-like data (95/299
mismatches on a 1e-6..1e6 series; all three of `test_period_boundary.c`'s sweep series are
two-decimal positives and blind to it). Do not reopen without pricing the period-1 arm it needs.

`RMA(x,n) != EMA(x,2n-1)` despite sharing `alpha`. incrediblecharts states the `1/14 == 27-day EMA`
smoothing-factor equivalence, but that is the factor only — the seed windows differ (`n` vs `2n-1`
bars). MEASURED at n=14: 0.159 absolute at the first common index, still 2.6e-10 after 270 bars.
RMA is therefore **not** expressible as a call to a shipped function and does **not** qualify for
the composite regtest category.

## Proposed API

```c
int TA_RMA_Lookback( int optInTimePeriod );   /* optInTimePeriod - 1
                                                 + TA_GetUnstablePeriod(<see open ruling 3>) */

TA_RetCode TA_RMA( int    startIdx,
                   int    endIdx,
                   const double inReal[],      /* plain real; no price bundle, no volume */
                   int    optInTimePeriod,     /* [1, 100000], default 30, suggested [1,200,1] */
                   int   *outBegIdx,
                   int   *outNBElement,
                   double outReal[] );         /* flags: [line] */
```

```yaml
name: RMA
camel_case: Rma
group: Overlap Studies
hint: Wilder's Smoothed Moving Average
flags: [overlap, stream, period1_identity]   # NOT start_dependent — RMA converges (see Notes)
```

Param block is byte-identical to EMA's (`input/ema/ema.yaml`).

Write `rma.c` **standalone**, not as a delegation to `ema_private`. Delegation would silently
inherit `ema_private`'s `TA_COMPATIBILITY_METASTOCK` branch (`input/ema/ema.c:100-113`), which is
retired for new functions, and would drag in the extra-`k`-param unguarded plumbing.

## References

- **ORIGINAL AUTHOR — J. Welles Wilder Jr., _New Concepts in Technical Trading Systems_, Trend
  Research, 1978.** Defines the smoothing used throughout RSI/ATR/ADX/DM.
- Achelis, _Technical Analysis from A to Z_, **p. 366** — the series Tulip's `atoz.txt:296` golden
  vector is transcribed from.
- thinkorswim WildersSmoothing —
  https://tlc.thinkorswim.com/center/reference/Tech-Indicators/studies-library/V-Z/WildersSmoothing
  ("smoothing factor of 1/length"; "The value for the first period is an SMA").
- Incredible Charts, Wilder Moving Average —
  https://www.incrediblecharts.com/indicators/wilder_moving_average.php (FETCHED: confirms
  `EMA% = 1/n` vs standard `2/(n+1)`; states the 27-day-EMA factor equivalence).
- TradingView Pine `ta.rma` — documented equivalent `alpha=1/length; sum := na(sum[1]) ?
  ta.sma(source,length) : alpha*source + (1-alpha)*nz(sum[1])`. **This is the proposed form,
  verbatim.**
- Tulip Indicators 0.9.2 `indicators/wilders.c:28-60`.
- pandas-ta-classic 0.6.52 `overlap/rma.py` (in-file comment: "SMA-seeded Wilder smoothing (matches
  TA-Lib)"); its own Sources list is the two links above. Kernel:
  `pandas/core/window/ewm.py:89`.

All sources agree on `alpha = 1/n` **and** on the SMA seed at bar `n-1`. There is no competing
seeding convention in mainstream use for this name. Note that generic `ewm(alpha=1/n).mean()`
recipes found on the web do *not* SMA-seed and diverge through the whole warm-up.

## Oracles

The external arms are **two families, not four** — group them by step form, not by alpha.

| Oracle | Status | Tolerance |
|---|---|---|
| **`TA_ATR` (in-tree)** | zero work, both sides ship here | **memcmp-exact** once both carry the shipped spelling. This is the only leg where bitwise is achievable *and* load-bearing — put the bitwise gate here. |
| `ta_tulip_serve` (`wilders`) | **arm exists** — `tulip_serve/server.c:394` | form-2 family. rel 1e-14 + an absolute companion |
| `ta4j_serve` (`MMAIndicator`) | arm to add | same form-2 kernel as Tulip; raw-seeded, so tail-convergent |
| `ta_trading_signals` (`WSMA`, and `RMA`) | arm to add, one line | `WSMA` is form 2; `RMA` is form 1 with pandas' kernel. Raw-seeded, tail-convergent |
| `ta_pandas_serve` (`ta.rma`) | arm to add — ~5 LOC, no build | form-1 family. **Drive it at our alpha, not `1/n`** — 4.1x tighter (see Formula) |
| Tulip checked-in vectors | already in-tree, zero work | printed precision, form-independent |

Every external arm is irreducibly ULP-scale away for reasons no spelling fixes, so they catch
structural bugs only — wrong seed window, anchor, alpha or sum order, all far above any tolerance.
That is what they are for; do not chase bitwise parity with them.

Tulip goldens, both REPRODUCED EXACTLY from the proposed formula:

```
untest.txt:483   wilders 5  ->  82.426 82.571 82.625 82.898 83.228 83.455 83.870
                                84.404 84.901 85.475 85.838                        (3dp, exact)
atoz.txt:296     wilders 5  ->  62.975 63.0675 63.054 63.1995 63.2408 63.2739
                 "#page 366"    63.1066 62.8540                                    (4dp, exact)
```

RMA is also the smoother behind pandas-ta's own rsi/atr/adx/cmo/dm/kdj, so that arm is reusable
later. `ta.rma(..., offset=)` MUST be left at 0 — TA-Lib has no contract for a shifted output.

## Verification plan

`test_rma.c` legs:

1. **The differential: `RMA(TRANGE(h,l,c), n) == TA_ATR(n)`, memcmp-exact**, over the reference
   series and gData, n in {3,14,30,100}, unstable period pinned 0. Strongest and cheapest leg, and
   the only gate that sees the coefficient spelling, the fusion and the fused operand ORDER — the
   last of which `fma.rs:150-152` documents as a live silent-failure class no static gate catches.
   Sabotage-prove it. **n=1 and n=2 are vacuous here** (identical coefficients under both
   spellings); never narrow this leg to them.
2. pandas golden on `TA_SREF_close_daily_ref_0_PRIV`, period 14, at rel 1e-12 **plus an absolute
   companion**, with `TA_SetUnstablePeriod(<unst id>, 0)` pinned — external oracles have no
   unstable-period concept. Drive the arm at our alpha (see Formula). Not bitwise: unreachable.
3. Tulip golden, same series, at rel 1e-14 — a tolerance arm under any form we can emit, since
   Tulip's `wilders` is unfused and its output is compiler-flag-dependent.
4. The two literal book vectors above as deterministic small-array checks.
5. **`RMA(x,1) == x` bitwise** — coefficients are exactly `(1,0)`, lookback 0, no period-1 branch
   needed (`period1_identity`, as SMA/EMA/WMA/DEMA/TEMA/TRIMA/KAMA/T3/HMA/VWMA/MA already carry).
   Pin `RMA(x,2)` too. Known hole, pre-existing and shared with SMA/TRIMA: with a real `fma`,
   `-0.0` comes back `+0.0` when the running value is positive, and all three
   `test_period_boundary.c` sweep series are strictly positive, so the sweep is blind to it. RMA is
   the first flagged member whose documented input domain is an arbitrary oscillator.
6. All-flat input returns the constant exactly and never NaN (#112). There is no data division
   (only `1.0/period`, `period>=1`), so no NaN/Inf path exists — this leg just pins that.
7. In-place aliasing `outReal == inReal` (write index `today-startIdx <= read index`, same
   argument as `ema_private`).
8. `server_verify("RMA", ...)` for cross-language bitwise.
9. `doRangeTestEx` in the CONVERGING class (MEASURED decay at period 14: seed error 0.168 at first
   output, 4.1e-3 after 50 bars, 6.2e-8 after 200).

Inherited free: `--codegen` all-4-language, `--xlang-hash` at zero tolerance (no transcendentals,
so no `CODEGEN_TRANSCENDENTAL[]` entry and no Java 1e-9 carve-out), the boundary/param sweep, ASan,
and `--fuzz-064` auto-skip via `serve_version.py` (new function).

## Decisions

1. **Coefficient spelling: beta-first, matching shipped ATR.** See Formula. Verified against the
   frozen legacy samples directly: alpha-first turns `LEGACY_TOL["ATR"]` red at period 19.
2. **Default period 30**, with the rest of the MA family. (Wilder's canonical is 14; pandas-ta uses
   10. Note both in `rma.md`.)
3. **Unstable-period id: APPEND `TA_FUNC_UNST_RMA`** to `enums.yaml`. The earlier claim that
   appending "shifts the `TA_FUNC_UNST_ALL` sentinel and breaks ABI" is false — `ALL` is pinned at
   65535 (`include/ta_defs.h:216`, and `backends/ta_defs.rs:98-130` says so). Reusing
   `TA_FUNC_UNST_UNUSED_12` is *not* the cheap move it looks: `test_internals.c:1151` names that
   constant in an explicitly append-only pin table, so the rename does not compile.
   Whichever id RMA gets, leg 1 only holds when RMA's unstable period matches ATR's — pin both to 0
   in the differential and document that the two knobs must agree.
4. **`TA_MAType_RMA` is a separate issue.** It would be value **13** (10 and 11 are the shipped
   `DISABLED`/`DEFAULT` pseudo-members, 12 is ZLEMA per proposal-drafts#32; the enum is
   append-only). Note that MAType is the route by which RMA reaches sign-crossing input.

## Must land in the same PR

- **`fma.rs::FUSING_INVENTORY` is an exact-set gate** — `cargo test` is red until `"rma"` is
  inserted between `"natr"` and `"sar"`.
- **Rewrite the "swapping them reddens nothing" comments** in `atr.c` and `natr.c`. They become
  false the day RMA ships: the memcmp differential turns the spelling into a gated contract
  (an alpha-first RMA mismatches shipped ATR on 215-288 of ~290 bars at n>=3).
- **The 1e-12 relative oracle tolerance needs an absolute companion.** It is 41x undersized on
  sign-crossing input (4.12e-11 at n=100 on `uniform(-100,100)`) — relative blow-up at a zero
  crossing, not a spelling effect. Pair them as `test_supertrend.c` and `test_kc.c` already do.
  Reachable in practice once `TA_MAType_RMA` exists.

## Optional, and it makes this reversible for free

A `ta_wilder_beta(period)` / `ta_wilder_alpha(beta)` pair in `ta_codegen/input/helpers/wilder.c`,
called from `atr.c`, `natr.c`, `supertrend.c` and `rma.c`, puts the spelling in one file instead of
four. Proven by running a full-tree `generate` in a sandbox: output bitwise identical, diff confined
to redundant parentheses.

**Do not extend the helper to the STEP.** `canonicalize_accumulator_add` (`backends/fma.rs:379`)
matches only a top-level `Add` and a `FuncCall` bypasses it, so a step helper written the natural
way (`return wa * x + wb * prev;`) silently re-emits every site as `fma(wAlpha, greatest, wBeta *
prevATR)` — 8 C and 6 Rust sites, batch and stream — changing 169/238 values at n=14 while reading
green on every existing gate.

## Notes

- **Streaming: free.** Same tier as `TA_EMA` — warm-up sum then a one-line scalar recurrence
  carrying a single `double`. `validate_streamable()` infers the same plan; `c_stream.rs` /
  `rust_stream.rs` / `java_stream.rs` emit from the existing match arms. Ship `flags: [..., stream]`
  from day one; unlike PVO there is no Composed-tier callee to wait on. Note the streaming tiers
  must carry `TA_FMA_MULTIVERSION` (TA-Lib/ta-lib#337) or the fused step costs a `call fma@PLT`
  per bar — which is the second reason this proposal waits for that issue.
- **Not `start_dependent`.** RMA converges; it is not the OBV/AD/SAR class (#127). Setting the flag
  would suppress real range-sweep coverage.
- Document in `rma.md`'s Implementation section that `TA_RMA(TRANGE(h,l,c), n)` **is** `TA_ATR(n)`
  bit-for-bit, and that the arithmetic form is Pine's and pandas'. Both are the questions users
  arrive with.
- Gotchas: run a FULL `cargo run -- generate` (a `--func=` filter skips the assembled `Core.java`);
  run `scripts/build.py servers`, not `build`, or the new-function gate coverage is silently stale;
  register `test_rma.c` in BOTH `CMakeLists.txt` and `src/tools/ta_regtest/Makefile.am` (the dist
  nightly is autotools — verify with `scripts/build.py check-source-lists`); add the CHANGELOG
  entry (forgotten for CMOU).

---
TS supported

**Comment by mario4tier** [comment]
Taking this one.

**Comment by mario4tier** [comment]
Landed on `dev`:

- `bbaae6c16` — TA_MAType_RMA = 13
- `f8ea6b68a` — TA_RMA itself, gated against TA_ATR by memcmp

Two corrections to this proposal, both measured:

**"Swapping them reddens nothing" (atr.c / natr.c) was already false before RMA shipped.** An alpha-first ATR moves 4.88e-15 from the frozen v0.6.4 samples against `LEGACY_TOL["ATR"]` 3e-15, and an alpha-first NATR 3.55e-15 against 2e-15. Both go red on the frozen comparison, independently of the new differential. Comments rewritten accordingly.

**`RMA(n)` and `EMA(2n-1)` do converge to bit-equality**, since `2/((2n-1)+1)` *is* `1/n` — only the seed window differs. Measured on the 1000-bar gData corpus at n=14: 0.518 at the first common bar, 4.7e-09 at +250, bit-equal from about +450. The proposal's "still 2.6e-10 after 270 bars" is a warm-up effect, not a permanent difference; `rma.md` says "differ through the warm-up" rather than "differ".

Two things the proposal got exactly right and that the sabotage runs confirmed:

- The **memcmp differential is the only in-tree gate that sees the coefficient spelling**. Under an alpha-first `rma.c`, pandas stays inside tolerance, Tulip stays inside tolerance, and both published vectors are still exact at printed precision. Only `TA_RMA(TA_TRANGE(h,l,c),n) == TA_ATR(n)` fails, at n=3 on the first bar.
- The **1e-12 relative oracle tolerance genuinely needs its absolute companion**. On a sign-crossing series, at the bar whose RMA is nearest zero, the error is ~1e-15 absolute but 1.06e-12 relative against pandas and 2.17e-12 against Tulip — a relative-only 1e-12 is red on a correct implementation.

The optional `ta_wilder_beta`/`ta_wilder_alpha` helper was not done: it is marked optional here, adds four input files to the diff, and buys no gate.

One item this proposal did not anticipate, recorded for the next MAType member: a wrong `ma_lookback` arm is invisible to every value and range comparison, because `ma()` forwards to a callee that clamps `startIdx` to its own lookback. The dispatch-parity leg compares the lookback tier directly.
